kandy-api/src/main/kotlin/org/jetbrains/kotlinx/kandy/dsl/internal/dataframe/DatasetBuilderImpl.kt

```
@@ -0,0 +1,91 @@
+/*
+* Copyright 2020-2023 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license.
+*/
+
+package org.jetbrains.kotlinx.kandy.dsl.internal.dataframe
+
+import org.jetbrains.kotlinx.dataframe.*
+import org.jetbrains.kotlinx.dataframe.DataFrame
+import org.jetbrains.kotlinx.dataframe.api.Infer
+import org.jetbrains.kotlinx.dataframe.api.*
+import org.jetbrains.kotlinx.dataframe.api.copy
+import org.jetbrains.kotlinx.dataframe.columns.ColumnAccessor
+import org.jetbrains.kotlinx.dataframe.columns.ColumnReference
+import org.jetbrains.kotlinx.kandy.dsl.internal.DatasetBuilder
+import org.jetbrains.kotlinx.kandy.ir.data.TableData
+
+/**
+ * [DatasetBuilder] implementation with DataFrame
+ */
+@PublishedApi
+internal abstract class DatasetBuilderImpl(
+    initialBuilder: DatasetBuilderImpl? = null
+): DatasetBuilder {
+    @PublishedApi
+    internal abstract val baseDataFrame: DataFrame<*>
+    private val referredColumns: MutableMap<String, String> = mutableMapOf()
+
+    @PublishedApi
+    internal var buffer: DataFrame<*> = initialBuilder?.buffer?.copy() ?: DataFrame.Empty
+
+    override fun rowsCount(): Int = if (baseDataFrame.isEmpty()) {
+        buffer.rowsCount()
+    } else baseDataFrame.rowsCount()
+
+    override fun takeColumn(name: String): String {
+        return referredColumns[name] ?: run {
+            val columnId = addColumn(
+                baseDataFrame.getColumnOrNull(name) ?: error("invalid column id: $name")
+            )
+            referredColumns[name] = columnId
+            name
+        }
+    }
+
+    fun takeColumn(column: ColumnAccessor<*>): String {
+        return takeColumn(column.name())
+    }
+
+    fun addColumn(column: ColumnReference<*>): String {
+        return when(column) {
+            is ColumnAccessor<*> -> takeColumn(column)
+            is DataColumn<*> -> addColumn(column)
+            else -> TODO()
```

I think need to add Exception or other types

---

That's interesting, I don't know why `ColumnReference` is not sealed in dataframe.

---

Okay, that doesn't concern us. In this case, you need to throw an exception